### PR TITLE
solves #591, refactoring beta release work

### PR DESCRIPTION
**Pull request by esteban-gs**
solves #591 
refactoring beta release commits.

### CODE REVIEW

File: Machete.Web/Controllers/Legacy/AccountController.cs [hunk 1]
```diff
@@ -213,8 +213,10 @@ public async Task<ActionResult> Register(RegisterViewModel model)
         {
             if (!ModelState.IsValid) return View(model);
             var newUserName = model.FirstName.Trim() + "." + model.LastName.Trim();
-            var dupeUser = await _userManager.FindByEmailAsync(model.Email);
-            if (dupeUser != null  && dupeUser.Email == model.Email) 
+            //Check for duplicate emails, if any, show error message
+            var allUsers = await _context.Users.ToListAsync();
```
Comment (esteban-gs) [inline Machete.Web/Controllers/Legacy/AccountController.cs]: Refactoring my previous commit - correct performance penalty

File: Machete.Web/Controllers/Legacy/AccountController.cs [hunk 2]
```diff
@@ -365,8 +367,10 @@ public async Task<ActionResult> Edit([Bind]EditUserViewModel model)
                 user.UserName = macheteUserName;
                 user.LoweredUserName = macheteUserName.ToLower();
                 //Check for duplicate emails, if any, show error message
-                var dupeUserCount = _context.Users.Count(u => u.Email == model.Email && u.Id != model.Id);
-                if (dupeUserCount > 0) 
+                var allUsers = await _context.Users.ToListAsync();
```
Comment (esteban-gs) [inline Machete.Web/Controllers/Legacy/AccountController.cs]: Refactoring my previous commit - correct performance penalty

File: Machete.Web/Controllers/Legacy/AccountController.cs [hunk 3]
```diff
@@ -407,11 +411,22 @@ private async Task<List<string>> TryChangePassword(MacheteUser user, string atte
 
             var remove = await _userManager.RemovePasswordAsync(user);
             if (!remove.Succeeded)
-                errors.Add("Something went wrong with your request. Contact an administrator for assistance.");
-
+            {
+                errors.Add("Something went wrong with your request. Contact an administrator for assistance.");
```
Comment (esteban-gs) [inline Machete.Web/Controllers/Legacy/AccountController.cs]: Adding userManager error strings to the list

File: Machete.Web/Controllers/Legacy/EmployerController.cs [hunk 4]
```diff
@@ -158,9 +158,11 @@ public async Task<JsonResult> Edit(int id, string userName)
             ModelState.ThrowIfInvalid();
 
             var employer = _serv.Get(id);
-
+            var employerOriginalEmail = employer.email;
```
Comment (esteban-gs) [inline Machete.Web/Controllers/Legacy/EmployerController.cs]: Mitigates #591

Comment (chaim1221) [reply Machete.Web/Controllers/Legacy/EmployerController.cs]: this certainly makes it impossible to edit, but does not explain to the user why. I see the TODO but it's better served as an issue that we paste the link to.

File: Machete.Web/Views/Shared/EditorTemplates/Employer.cshtml [hunk 5]
```diff
@@ -183,7 +183,7 @@
         </div>
         @* @Html.mUIDropDownYesNoFor(model => model.onlineSource, Model.def.yesnoSelectList(), new { tabindex = "20", id = idPrefix + "onlineSource" }) *@
         <div class="tb-field">
-            @Html.DropDownListFor(model => model.onlineSource, new SelectList(Model.def.yesnoSelectList(), "Value", "Text", Model), Shared.choose, new { tabindex = "20", id = idPrefix + "onlineSource" })
+            @Html.DropDownListFor(model => model.onlineSource, new SelectList(Model.def.yesnoSelectList(), "Value", "Text", Model), Shared.choose, new { tabindex = "20", id = idPrefix + "onlineSource",  @disabled = "disabled" })
```
Comment (chaim1221) [inline Machete.Web/Views/Shared/EditorTemplates/Employer.cshtml]: looks like maybe this is related to the above, but incomplete. I'd consider leaving it out for now.

File: Machete.Web/Controllers/Legacy/AccountController.cs (hunk 1 above)
Comment (esteban-gs) [reply Machete.Web/Controllers/Legacy/AccountController.cs]: Will adjust as recommended.

Comment (chaim1221) [reply Machete.Web/Controllers/Legacy/AccountController.cs]: `FindByEmailAsync` will only throw if null, just FYI. So you may want to check for null. If there are multiple matches, it returns the first. 
https://github.com/aspnet/Identity/blob/c7276ce2f76312ddd7fccad6e399da96b9f6fae1/src/EF/UserStore.cs#L634

File: Machete.Web/Controllers/Legacy/AccountController.cs [hunk 6]
```diff
@@ -364,14 +364,13 @@ public async Task<ActionResult> Edit([Bind]EditUserViewModel model)
                 var macheteUserName = model.FirstName.Trim() + "." + model.LastName.Trim();
                 user.UserName = macheteUserName;
                 user.LoweredUserName = macheteUserName.ToLower();
+                user.Email = model.Email.Trim();
                 //Check for duplicate emails, if any, show error message
-                var dupeUserCount = _context.Users.Count(u => u.Email == model.Email && u.Id != model.Id);
-                if (dupeUserCount > 0) 
+                if (_context.Users.Any(u => u.Email == user.Email && u.Id != user.Id))
```
Comment (esteban-gs) [inline Machete.Web/Controllers/Legacy/AccountController.cs]: Not able to use the `_userManager` here; cannot instantiate more than one user object.

File: Machete.Web/Controllers/Legacy/AccountController.cs [hunk 7]
```diff
@@ -407,11 +406,18 @@ private async Task<List<string>> TryChangePassword(MacheteUser user, string atte
 
             var remove = await _userManager.RemovePasswordAsync(user);
             if (!remove.Succeeded)
+            {
```
Comment (esteban-gs) [inline Machete.Web/Controllers/Legacy/AccountController.cs]: Reformatting happened from one commit to the next.

File: Machete.Web/Views/Shared/EditorTemplates/Employer.cshtml [hunk 8]
```diff
@@ -78,7 +78,14 @@
...
         <div class="tb-field">
-            @Html.TextBoxFor(model => model.email, new { tabindex = "8", id = idPrefix + "email" })
+            @if (!Convert.ToBoolean(ViewBag.onlineSource))
+            {
+                @Html.TextBoxFor(model => model.email, new { tabindex = "8", id = idPrefix + "email" })
+            }
+            @if (Convert.ToBoolean(ViewBag.onlineSource))
+            {
+                @Html.TextBoxFor(model => model.email, new { tabindex = "8", id = idPrefix + "email", @disabled = "disabled", @title="This Employer's Email is Managed by Machete Online Hiring Portal"})
+            }
```
Comment (esteban-gs) [inline Machete.Web/Views/Shared/EditorTemplates/Employer.cshtml]: #591

File: Machete.Web/Controllers/Legacy/AccountController.cs (hunk 6 above)
Comment (chaim1221) [reply Machete.Web/Controllers/Legacy/AccountController.cs]: I'm not sure what error message you got, but there's already an instantiated _userManager object. in any case if you're going to access the `_context` directly, don't do a non-indexed search over potentially thousands of users. 

```
var foo = _context.Users.Select(x => { x.Id, x.Email }).ToList();
var bar = foo.Where(y => y.Email == user.Email);
if (bar.Count() > 1) throw;
```
...etc.

File: Machete.Web/Views/Shared/EditorTemplates/Employer.cshtml [hunk 9]
```diff
@@ -78,7 +78,14 @@
             @Html.LabelFor(model => model.email)
         </div>
         <div class="tb-field">
-            @Html.TextBoxFor(model => model.email, new { tabindex = "8", id = idPrefix + "email" })
+            @if (!Convert.ToBoolean(ViewBag.onlineSource))
+            {
+                @Html.TextBoxFor(model => model.email, new { tabindex = "8", id = idPrefix + "email" })
+            }
+            @if (Convert.ToBoolean(ViewBag.onlineSource))
```
Comment (chaim1221) [inline Machete.Web/Views/Shared/EditorTemplates/Employer.cshtml]: use `model.onlinesource`. there should be no reason to use a weakly typed hashtable.

File: Machete.Web/Controllers/Legacy/AccountController.cs (hunk 6 above)
Comment (esteban-gs) [reply Machete.Web/Controllers/Legacy/AccountController.cs]: Well, the good thing is that my code here is unnecessary; now that the `userManager` errors are being displayed, there is no need for any of my dupeEmails code. 
image

Comment (esteban-gs) [reply Machete.Web/Controllers/Legacy/AccountController.cs]: Actually I spoke too soon. :)

File: Machete.Web/Controllers/Api/EmployersController.cs [hunk 10]
```diff
@@ -171,8 +171,9 @@ public ActionResult ProfilePut([FromBody]EmployerViewModel viewmodel)
                 newEmployer = true;
             }
             employer.onlineSigninID = UserSubject;
-            map.Map<EmployerViewModel, Domain.Employer>(viewmodel, employer);
             employer.email = UserEmail;
+            viewmodel.onlineSource = true;
+            map.Map<EmployerViewModel, Domain.Employer>(viewmodel, employer);
```
Comment (esteban-gs) [inline Machete.Web/Controllers/Api/EmployersController.cs]: The email here is sourced from the user store, so it's not in the viewmodel.

File: Machete.Web/Controllers/Legacy/AccountController.cs [hunk 11]
```diff
@@ -213,12 +213,6 @@ public async Task<ActionResult> Register(RegisterViewModel model)
         {
             if (!ModelState.IsValid) return View(model);
             var newUserName = model.FirstName.Trim() + "." + model.LastName.Trim();
-            var dupeUser = await _userManager.FindByEmailAsync(model.Email);
-            if (dupeUser != null  && dupeUser.Email == model.Email) 
-            {
-                ModelState.AddModelError("", ValidationStrings.dupeEmail);
-                return View(model);
-            }   
             var user = new MacheteUser { UserName = newUserName, LoweredUserName = newUserName.ToLower(), ApplicationId = GetApplicationId(), Email = model.Email.Trim(), LoweredEmail = model.Email.Trim() };
```
Comment (esteban-gs) [inline Machete.Web/Controllers/Legacy/AccountController.cs]: All this is handled by the user manager on Register, errors still get displayed as feedback.

**Pull request merged**